storage/range.go

```diff
@@ -789,32 +786,71 @@ func (r *Range) processRaftCommand(idKey cmdIDKey, index uint64, raftCmd proto.I
 // underlying state machine (i.e. the engine).
 // When certain critical operations fail, a replicaCorruptionError may be
 // returned and must be handled by the caller.
-func (r *Range) applyRaftCommand(ctx context.Context, index uint64, originNode proto.RaftNodeID, args proto.Request, reply proto.Response) (rErr error) {
+func (r *Range) applyRaftCommand(ctx context.Context, index uint64, originNode proto.RaftNodeID, args proto.Request, reply proto.Response) error {
 	if index <= 0 {
 		log.Fatalc(ctx, "raft command index is <= 0")
 	}
 
-	committed := false
-	// The very last thing we do before returning is move the applied index
-	// forward, unless that has already happened as part of a successfully
-	// committed batch.
-	defer func() {
-		if !committed {
-			// We didn't commit the batch, but advance the last applied index nonetheless.
-			if err := setAppliedIndex(r.rm.Engine(), r.Desc().RaftID, index); err != nil {
-				rErr = newReplicaCorruptionError(
-					util.Errorf("could not advance applied index"), err, rErr)
-				return
+	// If we have an out of order index, there's corruption. No sense in trying
+	// to update anything or run the command. Simply return a corruption error.
+	if oldIndex := atomic.LoadUint64(&r.appliedIndex); oldIndex >= index {
+		return newReplicaCorruptionError(util.Errorf("applied index moved backwards: %d >= %d", oldIndex, index))
+	}
+
+	// Call the helper, which returns a batch containing data written
+	// during command execution and any associated error.
+	ms := engine.MVCCStats{}
+	batch, rErr := r.applyRaftCommandHelper(ctx, index, originNode, args, reply, &ms)
+	defer batch.Close()
+
+	// Advance the last applied index and commit the batch.
+	if err := setAppliedIndex(batch, r.Desc().RaftID, index); err != nil {
+		log.Fatalc(ctx, "setting applied index in a batch should never fail: %s", err)
+	}
+	if err := batch.Commit(); err != nil {
+		rErr = newReplicaCorruptionError(util.Errorf("could not commit batch"), err, rErr)
+	}
+
+	// On successful write commands, flush to event feed, and handle other
+	// write-related triggers including splitting and config gossip updates.
+	if rErr == nil && proto.IsWrite(args) {
+		// Publish update to event feed.
+		r.rm.EventFeed().updateRange(r, args.Method(), &ms)
+		// If the commit succeeded, potentially add range to split queue.
+		r.maybeAddToSplitQueue()
+		// Maybe update gossip configs on a put.
+		switch args.(type) {
+		case *proto.PutRequest, *proto.DeleteRequest, *proto.DeleteRangeRequest:
+			if key := args.Header().Key; key.Less(keys.SystemMax) {
+				// We hold the lock already.
+				r.maybeGossipConfigsLocked(func(configPrefix proto.Key) bool {
+					return bytes.HasPrefix(key, configPrefix)
+				})
 			}
-			atomic.StoreUint64(&r.appliedIndex, index)
 		}
-	}()
+	}
+
+	// Update cached appliedIndex on anything but a replica corruption error.
+	if _, ok := rErr.(*replicaCorruptionError); !ok {
```

Comment:
Does it make sense to be stingy here? After all, if the corruption happened during `batch.Commit()`, updating might actually be the more consistent thing to do. I would just by convention say that the appliedIndex is invalid once you have corruption, and that you just always update it via `defer()` so that this cached variable reliably reflects where things went wrong.